File: common/src/main/java/earth/terrarium/techarium/common/blockentities/base/DeployableBlockEntity.java
package earth.terrarium.techarium.common.blockentities.base;

import earth.terrarium.techarium.common.registry.ModSoundEvents;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.NotNull;

public abstract class DeployableBlockEntity extends ContainerMachineBlockEntity {
    private int deploymentTicks = 0;

    public DeployableBlockEntity(BlockPos pos, BlockState state, int containerSize) {
        super(pos, state, containerSize);
    }

    @Override
    protected void saveAdditional(@NotNull CompoundTag tag) {
        super.saveAdditional(tag);
        tag.putInt("DeploymentTicks", deploymentTicks);
    }

    @Override
    public void load(@NotNull CompoundTag tag) {
        super.load(tag);
        deploymentTicks = tag.getInt("DeploymentTicks");
    }

    @Override
    public void firstTick(Level level, BlockPos pos, BlockState state) {
        super.firstTick(level, pos, state);
        if (level.isClientSide()) return;
        if (deploymentTicks < getDeploymentTime()) {
            level.playSound(null, pos, getDeploySound(), SoundSource.BLOCKS, 0.5f, 1.0f);
        }
    }

    @Override
    public void internalServerTick(ServerLevel level, long time, BlockState state, BlockPos pos) {
        super.internalServerTick(level, time, state, pos);
        if (deploymentTicks < getPartialDeploymentTime()) {
            deploymentTicks++;
        } else if (deploymentTicks < getDeploymentTime()) {
            level.setBlockAndUpdate(pos, state.setValue(DeployableBlock.DEPLOYED, DeployableBlock.DeployState.PARTIALLY_DEPLOYED));
            deploymentTicks++;
        } else {
            level.setBlockAndUpdate(pos, state.setValue(DeployableBlock.DEPLOYED, DeployableBlock.DeployState.DEPLOYED));
        }
    }

    public boolean isDeployed() {
        return getBlockState().getValue(DeployableBlock.DEPLOYED).isDeployed();
    }

    public SoundEvent getDeploySound() {
        return ModSoundEvents.BOTARIUM_DEPLOY.get();
    }

    public abstract int getPartialDeploymentTime();

    public abstract int getDeploymentTime();
}

File: src/main/kotlin/earth/terrarium/techarium/common/blocks/entities/DeployableBlockEntity.kt
package earth.terrarium.techarium.common.blocks.entities

import earth.terrarium.techarium.common.blocks.base.TechariumBlockStates.DEPLOYED
import earth.terrarium.techarium.common.blocks.base.Tickable
import net.minecraft.core.BlockPos
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import software.bernie.geckolib.animatable.GeoBlockEntity
import software.bernie.geckolib.animatable.instance.AnimatableInstanceCache
import software.bernie.geckolib.util.GeckoLibUtil

abstract class DeployableBlockEntity(
    type: BlockEntityType<*>,
    pos: BlockPos, state: BlockState
) : ComponentBlockEntity(type, pos, state), GeoBlockEntity, Tickable {

    private val cache = GeckoLibUtil.createInstanceCache(this)
    var ticks = 0
        private set

    protected val isDeployed: Boolean
        get() = this.blockState.hasProperty(DEPLOYED) && this.blockState.getValue(DEPLOYED)

    final override fun getAnimatableInstanceCache(): AnimatableInstanceCache = cache

    final override fun tick(level: Level, pos: BlockPos, state: BlockState) {
        ticks++
        super.tick(level, pos, state)
    }
}